Refactor DynacastQuality & MediaLossProxy into separate modules

boks1971 opened the pull request:
- These get used from top level `MediaTrack`.

boks1971 commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,234 @@
+	Logger             logger.Logger
+}
+
+type DynacastManager struct {
```
A lot of this PR is just refactoring. There is a `DynacastQuality` (another new module 👇 ) module per receiver in a multi-codec receiver. This module `DynacastManager` aggregates all the codecs of a receiver. Splitting these out into smaller modules as readability in `MediaTrackSubscriptions` was getting hard. And also, for relay track, these are maintained by relay manager and hence made sense to split them out.

boks1971 commented on pkg/rtc/mediatrack.go:
```diff
@@ -28,6 +30,9 @@ type MediaTrack struct {
 	*MediaTrackReceiver
+	*MediaLossProxy
+
+	dynacastManager *DynacastManager
```
These got moved out of `MediaTrackReceiver` and `MediaTrackSubscriptions` into here so that local media track is managed here and remote media track is managed in relay manager.

davidzhao replied:
great, I think this is cleaner!

boks1971 commented on pkg/rtc/mediatrackreceiver.go:
```diff
@@ -5,7 +5,6 @@ import (
 	"sort"
 	"strings"
 	"sync"
-	"time"
```
Changes here are to move out the dynacast and media loss proxy stuff into their own modules and into top level `MediaTrack`.

boks1971 commented on pkg/rtc/mediatracksubscriptions.go:
```diff
@@ -6,7 +6,6 @@ import (
 	"sync"
 	"time"
 
-	"github.com/bep/debounce"
```
Once again changes here are move out the dynacast stuff and media loss proxy into their own modules.

boks1971 commented on pkg/rtc/participant.go:
```diff
@@ -2328,3 +2329,25 @@ func (p *ParticipantImpl) ClearInProgressAndProcessSubscriptionRequestsQueue(tra
 	go p.ProcessSubscriptionRequestsQueue(trackID)
 }
+
+func (p *ParticipantImpl) UpdateSubscribedQuality(nodeID livekit.NodeID, trackID livekit.TrackID, maxQualities []types.SubscribedCodecQuality) error {
```
These are moved to `LocalParticipant` as for relay track they do not go through `RemoteParticipant`, but directly to the relay manager.

boks1971 commented on pkg/rtc/types/interfaces.go:
```diff
@@ -348,23 +347,25 @@ type MediaTrack interface {
 	GetQualityForDimension(width, height uint32) livekit.VideoQuality
 
-	NotifySubscriberNodeMaxQuality(nodeID livekit.NodeID, qualites []SubscribedCodecQuality)
-	NotifySubscriberNodeMediaLoss(nodeID livekit.NodeID, fractionalLoss uint8)
```
These are moving from `MediaTrack` -> `LocalMediaTrack` as once again relay track will directly go to the relay maanger.

boks1971 commented on pkg/sfu/downtrack.go:
```diff
@@ -306,6 +306,9 @@ func (d *DownTrack) Bind(t webrtc.TrackLocalContext) (webrtc.RTPCodecParameters,
 
+	if d.onMaxLayerChanged != nil {
+		d.onMaxLayerChanged(d, d.MaxLayers().Spatial)
+	}
```
`MediaTrackSubscriptions` was setting max layer to `HIGH` on adding a down track. But, that was hanging without much context. So, calling this from down track on bind so that it comes from the subscriber.

davidzhao commented on pkg/rtc/mediatrack.go:
```diff
@@ -281,3 +319,26 @@ func (t *MediaTrack) OnMaxLayerChange(maxLayer int32) {
 	t.params.Telemetry.TrackPublishedUpdate(context.Background(), t.PublisherID(), ti)
 }
+
+func (t *MediaTrack) Restart() {
```
when would a MediaTrack be restarted?

boks1971 replied:
Happens on an ICERestart.

davidzhao commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,265 @@
+	Logger             logger.Logger
+}
+
+type DynacastManager struct {
```
😍 this is fantastic!  do you think we should move the test cases over from MediaTrackTest to here too?

boks1971 replied:
Will move them.

davidzhao commented on pkg/rtc/uptrackmanager.go:
```diff
@@ -60,7 +60,7 @@ func (u *UpTrackManager) Start() {
 func (u *UpTrackManager) Restart() {
 	for _, t := range u.GetPublishedTracks() {
-		t.Restart()
+		t.(types.LocalMediaTrack).Restart()
```
would there be potentially casting errors? i.e. a regular MediaTrack in the list of published tracks?

boks1971 replied:
I think this can be moved to `LocalParticipant` (where tracks will be of type `LocalMediaTrack` so the casting will not look odd). Will move it.

davidzhao reviewed (approved):
Awesome work!

cnderrauber commented on pkg/rtc/dynacastquality.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func (d *DynacastQuality) Restart() {
+	d.reset()
```
can we merge restart and start method? they are same

boks1971 replied:
From an API perspective, two things made sense. As you can `Restart` without calling `Stop`. So, made a separate API.

cnderrauber commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,265 @@
+			})
+		}
+	}
+	if d.onSubscribedMaxQualityChange != nil {
```
can do this check before calculate subscribedCodecs

boks1971 replied:
Will change.

cnderrauber reviewed (approved):
Great work!!

boks1971 commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,265 @@
+	}
+}
+
+func (d *DynacastManager) ForceUpdate() {
```
Will add comments.

boks1971 commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,265 @@
+	d.update(true)
+}
+
+func (d *DynacastManager) ForceQuality(quality livekit.VideoQuality) {
```
This is used in relay path when the relay track is moved to closing state. Will add comments. Relay comments don't make a lot of sense in OSS, but will add generic comments.

boks1971 commented on pkg/rtc/dynacastmanager.go:
```diff
@@ -0,0 +1,265 @@
+		return
+	}
+
+	subscribedCodec := make([]*livekit.SubscribedCodec, 0, len(d.committedMaxSubscribedQuality))
```
Will change.

boks1971 commented on pkg/rtc/dynacastquality.go:
```diff
@@ -0,0 +1,162 @@
+	Logger             logger.Logger
+}
+
+// DybacastQuality manages max subscribed quality of a media track
```
Will change.

boks1971 commented:
@davidzhao @cnderrauber Addressing PR comments in this commit - https://github.com/livekit/livekit/pull/894/commits/7dcfcdfc5fe25809fbe8aa2348ee6cae94eee3d4. Please let me know if I missed something.

boks1971 commented:
Going to merge this, but if you have further comments, please let me know and I will address in a subsequent PR.
